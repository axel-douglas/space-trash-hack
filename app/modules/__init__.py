# app/modules/__init__.py
"""
Exportes livianos para la app Streamlit.

Notas:
- NO importamos `model_training` al cargar el paquete (lazy import).
- Re-exportamos solo runtime estable (IO, generator, ML registry).
- Si se quiere entrenar desde la UI/CLI: usar `get_train_and_save()`.
"""

from __future__ import annotations

# --- Runtime estable (ligero) ---
from importlib import import_module
from typing import Any

from .ml_models import MODEL_REGISTRY, ModelRegistry, PredictionResult
from .io import (
    load_waste_df,
    save_waste_df,
    load_targets,
    load_process_catalog,
)

__all__ = [
    # IO
    "load_waste_df",
    "save_waste_df",
    "load_targets",
    "load_process_catalog",
    # Generación
    "generate_candidates",
    "PredProps",
    "rank_candidates",
    "score_recipe",
    "derive_auxiliary_signals",
    "suggest_next_candidates",
    "Acquisition",
    # ML
    "MODEL_REGISTRY",
    "ModelRegistry",
    "PredictionResult",
    # Entrenamiento (lazy)
    "get_train_and_save",
]

def get_train_and_save():
    """
    Import diferido del pipeline de entrenamiento.
    Evita errores/sobrecargas cuando solo se usa la app.
    """
    try:
        from .model_training import train_and_save  # import tardío
        return train_and_save
    except Exception as exc:
        def _stub(*_a, **_kw):
            raise RuntimeError(f"Training pipeline no disponible: {exc}")
        return _stub


_LAZY_MODULES = {
    "generator": {
        "generate_candidates",
        "PredProps",
    },
    "ranking": {
        "rank_candidates",
        "score_recipe",
        "derive_auxiliary_signals",
    },
    "active_learning": {
        "suggest_next_candidates",
        "Acquisition",
    },
}


def __getattr__(name: str) -> Any:
    for module_name, symbols in _LAZY_MODULES.items():
        if name in symbols:
            module = import_module(f".{module_name}", __name__)
            value = getattr(module, name)
            globals()[name] = value
            return value
    raise AttributeError(f"module {__name__!r} has no attribute {name!r}")
