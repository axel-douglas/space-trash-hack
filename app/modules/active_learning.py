"""Active learning helpers to prioritise the next recipes to validate.

The module operates on the candidate payloads generated by
``app.modules.generator.generate_candidates``.  The helpers score each
candidate based on uncertainty or expected improvement to aid operators when
selecting the next physical or simulated experiment.
"""

from __future__ import annotations

from dataclasses import dataclass
from statistics import mean
from typing import Any, Iterable, Mapping, MutableMapping, Sequence

TARGET_COLUMNS = ("rigidez", "estanqueidad", "energy_kwh", "water_l", "crew_min")


@dataclass(slots=True)
class Acquisition:
    """Representation of a candidate selected for active learning."""

    recipe_id: str
    process_id: str
    priority: float
    strategy: str
    metrics: MutableMapping[str, float]
    candidate: Mapping[str, Any]


def _as_mapping(payload: Any) -> Mapping[str, Any]:
    if isinstance(payload, Mapping):
        return payload
    if hasattr(payload, "_asdict"):
        try:
            return payload._asdict()  # type: ignore[attr-defined]
        except Exception:  # pragma: no cover - defensive
            return {}
    if hasattr(payload, "__dict__"):
        return getattr(payload, "__dict__")
    return {}


def _extract_confidence(candidate: Mapping[str, Any]) -> Mapping[str, tuple[float, float]]:
    payload = candidate.get("confidence_interval")
    if isinstance(payload, Mapping):
        try:
            return {
                str(name): (float(bounds[0]), float(bounds[1]))
                for name, bounds in payload.items()
            }
        except Exception:  # pragma: no cover - defensive
            return {}

    props = candidate.get("props")
    if props is not None and hasattr(props, "confidence_interval"):
        mapping = getattr(props, "confidence_interval")
        if isinstance(mapping, Mapping):
            return {
                str(name): (float(bounds[0]), float(bounds[1]))
                for name, bounds in mapping.items()
            }

    features = candidate.get("features")
    if isinstance(features, Mapping):
        mapping = features.get("confidence_interval")
        if isinstance(mapping, Mapping):
            try:
                return {
                    str(name): (float(bounds[0]), float(bounds[1]))
                    for name, bounds in mapping.items()
                }
            except Exception:  # pragma: no cover - defensive
                return {}

    return {}


def _extract_uncertainty(candidate: Mapping[str, Any]) -> Mapping[str, float]:
    payload = candidate.get("uncertainty")
    if isinstance(payload, Mapping):
        try:
            return {str(name): float(value) for name, value in payload.items()}
        except Exception:  # pragma: no cover - defensive
            return {}

    props = candidate.get("props")
    if props is not None and hasattr(props, "uncertainty"):
        mapping = getattr(props, "uncertainty")
        if isinstance(mapping, Mapping):
            return {str(name): float(value) for name, value in mapping.items()}

    features = candidate.get("features")
    if isinstance(features, Mapping):
        mapping = features.get("uncertainty")
        if isinstance(mapping, Mapping):
            try:
                return {str(name): float(value) for name, value in mapping.items()}
            except Exception:  # pragma: no cover - defensive
                return {}

    return {}


def _confidence_width(bounds: tuple[float, float]) -> float:
    low, high = bounds
    width = float(high) - float(low)
    return max(width, 0.0)


def _uncertainty_score(candidate: Mapping[str, Any], targets: Iterable[str]) -> float:
    targets = list(targets)
    ci_map = _extract_confidence(candidate)
    widths = [_confidence_width(ci_map[name]) for name in targets if name in ci_map]

    if widths:
        return float(mean(widths))

    unc_map = _extract_uncertainty(candidate)
    values = [abs(float(unc_map[name])) for name in targets if name in unc_map]
    if values:
        return float(mean(values))

    return 0.0


def _best_score(observations: Sequence[Mapping[str, Any]] | None) -> float:
    if not observations:
        return 0.0

    best = None
    for obs in observations:
        try:
            score = float(obs.get("score", 0.0))
        except Exception:  # pragma: no cover - defensive
            continue
        if best is None or score > best:
            best = score

    return float(best) if best is not None else 0.0


def _candidate_score(candidate: Mapping[str, Any]) -> float:
    if "score" in candidate:
        try:
            return float(candidate["score"])
        except Exception:  # pragma: no cover - defensive
            return 0.0

    breakdown = candidate.get("score_breakdown")
    if isinstance(breakdown, Mapping) and "total" in breakdown:
        try:
            return float(breakdown["total"])
        except Exception:  # pragma: no cover - defensive
            return 0.0

    props = candidate.get("props")
    if props is not None and hasattr(props, "to_targets"):
        try:
            targets = props.to_targets()
            if isinstance(targets, Mapping) and "rigidez" in targets:
                return float(targets["rigidez"])
        except Exception:  # pragma: no cover - defensive
            return 0.0

    return 0.0


def _candidate_ids(candidate: Mapping[str, Any]) -> tuple[str, str]:
    recipe_id = None
    process_id = None

    features = candidate.get("features")
    if isinstance(features, Mapping):
        recipe_id = features.get("recipe_id")
        process_id = features.get("process_id")

    if recipe_id is None:
        recipe_id = candidate.get("recipe_id")
    if process_id is None:
        process_id = candidate.get("process_id")

    if recipe_id is None and hasattr(candidate.get("props"), "source"):
        recipe_id = getattr(candidate.get("props"), "source")

    return str(recipe_id or ""), str(process_id or "")


def suggest_next_candidates(
    candidates: Sequence[Mapping[str, Any]],
    *,
    strategy: str = "uncertainty",
    top_n: int = 5,
    observations: Sequence[Mapping[str, Any]] | None = None,
    targets: Iterable[str] = TARGET_COLUMNS,
    exploration_bonus: float = 0.1,
) -> list[Acquisition]:
    """Rank candidates according to the requested active learning strategy."""

    if not candidates:
        return []

    targets = list(targets)
    ranked: list[Acquisition] = []

    best_score = _best_score(observations)
    for candidate in candidates:
        recipe_id, process_id = _candidate_ids(candidate)
        metrics: MutableMapping[str, float] = {}

        if strategy == "uncertainty":
            priority = _uncertainty_score(candidate, targets)
            metrics["uncertainty"] = priority
        elif strategy == "expected_improvement":
            predicted = _candidate_score(candidate)
            uncertainty = _uncertainty_score(candidate, targets)
            improvement = max(0.0, predicted - best_score)
            priority = improvement + exploration_bonus * uncertainty
            metrics.update(
                {
                    "predicted_score": predicted,
                    "baseline_score": best_score,
                    "improvement": improvement,
                    "uncertainty": uncertainty,
                }
            )
        else:  # pragma: no cover - defensive against typos
            raise ValueError(f"Unknown active learning strategy: {strategy}")

        ranked.append(
            Acquisition(
                recipe_id=recipe_id,
                process_id=process_id,
                priority=float(priority),
                strategy=strategy,
                metrics=metrics,
                candidate=candidate,
            )
        )

    ranked.sort(key=lambda item: item.priority, reverse=True)
    return ranked[: max(top_n, 0)]


__all__ = ["Acquisition", "suggest_next_candidates", "TARGET_COLUMNS"]

