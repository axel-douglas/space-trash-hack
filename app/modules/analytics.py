import numpy as np
import pandas as pd


def pareto_front(
    df: pd.DataFrame,
    minimize_cols=("Energía (kWh)", "Agua (L)", "Crew (min)"),
    maximize_cols=("Score",),
):
    """Devuelve los índices de las filas no dominadas (frente de Pareto).

    La función mantiene la interfaz existente (lista de índices del ``DataFrame``
    original) pero ahora opera directamente sobre ``numpy`` sin crear columnas
    auxiliares.  Se construye una matriz con las columnas a minimizar y el
    negativo de las columnas a maximizar, y se utilizan operaciones vectorizadas
    para evaluar la dominancia entre todas las combinaciones de filas.  Esta
    versión evita bucles en Python y reduce el coste de copiar datos,
    proporcionando un mejor rendimiento en ``DataFrames`` grandes.
    """

    metrics = []
    if minimize_cols:
        metrics.append(df.loc[:, list(minimize_cols)].to_numpy())
    if maximize_cols:
        metrics.append(-df.loc[:, list(maximize_cols)].to_numpy())

    if not metrics:
        return df.index.tolist()

    X = np.hstack(metrics) if len(metrics) > 1 else metrics[0]

    less_equal = X[:, None, :] <= X[None, :, :]
    strictly_less = X[:, None, :] < X[None, :, :]
    dominates = np.all(less_equal, axis=2) & np.any(strictly_less, axis=2)
    is_dominated = dominates.any(axis=0)
    front_idx = np.nonzero(~is_dominated)[0]
    return df.index[front_idx].tolist()
