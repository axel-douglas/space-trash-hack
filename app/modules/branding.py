# --- path guard universal ---
import sys, pathlib
_here = pathlib.Path(__file__).resolve()
p = _here.parent
while p.name != "app" and p.parent != p:
    p = p.parent
repo_root = p.parent if p.name == "app" else _here.parent
if str(repo_root) not in sys.path:
    sys.path.insert(0, str(repo_root))
# --------------------------------

from pathlib import Path
import streamlit as st

PRIMARY   = "#0A84FF"
PRIMARY_2 = "#3AA6FF"
OK        = "#00E38C"
WARN      = "#FFB020"
RISK      = "#FF5C5C"
CARD      = "#171B2C"
BORDER    = "#2A2F45"

STATIC_DIR = (repo_root / "app" / "static")
LOGO_PATH  = STATIC_DIR / "logo_rexai.svg"

def inject_branding():
    STATIC_DIR.mkdir(parents=True, exist_ok=True)
    st.markdown(
        f"""
        <style>
          .stApp {{
            background: linear-gradient(180deg,#0B0E1C 0%,#0B0E1C 60%,#0A0D1A 100%);
            color:#E8ECF4;
          }}
          .rex-header {{ display:flex; gap:16px; align-items:center; padding:6px 8px 2px 4px; }}
          .rex-title {{ font-weight:800; letter-spacing:.3px; font-size:22px; margin:0; }}
          .rex-sub {{ color:#A9B2C8; margin-top:2px; font-size:13px; }}
          .rex-card {{
             background:{CARD}; border:1px solid {BORDER};
             border-radius:16px; padding:16px; box-shadow:0 8px 24px rgba(0,0,0,.22);
          }}
          .rex-chip {{ display:inline-flex; align-items:center; gap:8px; border-radius:999px;
             padding:4px 10px; font-size:12px; background:#1E2438; border:1px solid {BORDER};
             margin-right:8px; margin-top:6px;
          }}
          [data-testid="stSidebar"] > div:first-child {{
            background:#0D1122; border-right:1px solid {BORDER};
          }}
          .rex-logo svg {{ width:160px; height:auto; color:{PRIMARY}; display:block; }}
        </style>
        """,
        unsafe_allow_html=True,
    )

    svg = ""
    try:
        if LOGPATH := (LOGO_PATH if LOGO_PATH.exists() else None):
            svg = LOGPATH.read_text(encoding="utf-8")
    except Exception:
        pass

    st.markdown(
        f"""
        <div class="rex-header">
          <div class="rex-logo">{svg}</div>
          <div>
            <div class="rex-title">REX-AI Mars</div>
            <div class="rex-sub">Recycling & Experimentation eXpert â€” Jezero Base</div>
          </div>
        </div>
        """,
        unsafe_allow_html=True,
    )
