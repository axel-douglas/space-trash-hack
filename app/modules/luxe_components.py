"""High fidelity UI components for the Rex-AI Streamlit app.

This module centralises visually rich components (hero, metrics, glass cards)
so that pages can consume declarative helpers instead of raw HTML blocks.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Mapping, Sequence

import streamlit as st


# ---------------------------------------------------------------------------
# CSS bootstrap
# ---------------------------------------------------------------------------
_LUXE_CSS = """
<style>
:root {
  --luxe-surface: rgba(12, 17, 27, 0.78);
  --luxe-border: rgba(148, 163, 184, 0.32);
  --luxe-border-strong: rgba(148, 163, 184, 0.48);
  --luxe-ink: #e9f0ff;
  --luxe-muted: rgba(226, 232, 240, 0.76);
  --luxe-accent: #60a5fa;
  --luxe-positive: #34d399;
  --luxe-warning: #f59e0b;
  --luxe-danger: #f87171;
}

@keyframes heroGlow {
  0% { box-shadow: 0 25px 70px rgba(96, 165, 250, 0.24); }
  50% { box-shadow: 0 35px 110px rgba(56, 189, 248, 0.45); }
  100% { box-shadow: 0 25px 70px rgba(96, 165, 250, 0.24); }
}

@keyframes parallaxDrift {
  0% { transform: translate3d(0, 0, 0) scale(1); opacity: 0.45; }
  50% { transform: translate3d(12px, -10px, 0) scale(1.05); opacity: 0.75; }
  100% { transform: translate3d(0, 0, 0) scale(1); opacity: 0.45; }
}

@keyframes sparkleShift {
  0% { background-position: 0% 50%; }
  50% { background-position: 100% 50%; }
  100% { background-position: 0% 50%; }
}

.luxe-hero {
  position: relative;
  overflow: hidden;
  border-radius: 32px;
  border: 1px solid var(--luxe-border);
  background: var(--hero-gradient, linear-gradient(135deg, rgba(59, 130, 246, 0.16), rgba(14, 165, 233, 0.06)));
  padding: var(--hero-padding, 2.6rem 3.1rem);
  color: var(--luxe-ink);
  isolation: isolate;
  backdrop-filter: blur(18px);
  box-shadow: 0 24px 60px rgba(8, 15, 35, 0.45);
  animation: heroGlow 14s ease-in-out infinite;
}

.luxe-hero::after {
  content: "";
  position: absolute;
  inset: -30%;
  background: radial-gradient(circle at 20% 20%, var(--hero-glow, rgba(96, 165, 250, 0.4)), transparent 62%);
  z-index: -2;
  filter: blur(2px);
  animation: sparkleShift 28s ease infinite;
}

.luxe-hero__layer {
  position: absolute;
  pointer-events: none;
  opacity: 0.55;
  font-size: var(--layer-size, 4rem);
  animation: parallaxDrift var(--layer-speed, 18s) ease-in-out infinite;
  mix-blend-mode: screen;
}

.luxe-hero__content {
  position: relative;
  z-index: 1;
  display: flex;
  flex-direction: column;
  gap: 0.85rem;
}

.luxe-hero__icon {
  font-size: 2.2rem;
  align-self: flex-start;
  filter: drop-shadow(0 0 10px rgba(148, 163, 184, 0.45));
}

.luxe-hero h1 {
  font-size: clamp(2.15rem, 4vw, 2.9rem);
  margin: 0;
  letter-spacing: 0.015em;
}

.luxe-hero p {
  margin: 0;
  color: var(--luxe-muted);
  font-size: 1.05rem;
  max-width: 46rem;
}

.luxe-chip-row {
  display: flex;
  gap: var(--chip-gap, 0.6rem);
  flex-wrap: wrap;
  margin-top: var(--chip-margin-top, 0.6rem);
}

.luxe-chip {
  border-radius: 999px;
  padding: var(--chip-padding, 0.38rem 0.9rem);
  border: 1px solid var(--chip-border, rgba(148, 163, 184, 0.35));
  background: var(--chip-bg, rgba(15, 23, 42, 0.55));
  color: var(--chip-ink, var(--luxe-ink));
  font-size: var(--chip-size, 0.82rem);
  font-weight: 600;
  display: inline-flex;
  align-items: center;
  gap: 0.35rem;
  backdrop-filter: blur(6px);
}

.luxe-chip[data-tone="accent"] { background: rgba(96, 165, 250, 0.16); border-color: rgba(125, 211, 252, 0.45); }
.luxe-chip[data-tone="info"] { background: rgba(14, 165, 233, 0.16); border-color: rgba(56, 189, 248, 0.5); }
.luxe-chip[data-tone="positive"] { background: rgba(52, 211, 153, 0.14); border-color: rgba(52, 211, 153, 0.45); }
.luxe-chip[data-tone="warning"] { background: rgba(245, 158, 11, 0.12); border-color: rgba(245, 158, 11, 0.45); }

.luxe-metric-galaxy {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(var(--metric-min, 13rem), 1fr));
  gap: var(--metric-gap, 1rem);
}

.luxe-metric {
  border-radius: 22px;
  border: 1px solid var(--luxe-border);
  background: rgba(13, 17, 23, 0.76);
  padding: var(--metric-padding, 1.2rem 1.4rem);
  box-shadow: inset 0 0 0 1px rgba(148, 163, 184, 0.1), 0 18px 40px rgba(8, 15, 35, 0.38);
  display: flex;
  flex-direction: column;
  gap: 0.35rem;
  position: relative;
  overflow: hidden;
}

.luxe-metric[data-glow="true"]::after {
  content: "";
  position: absolute;
  inset: 0;
  background: radial-gradient(circle at 15% 20%, rgba(96, 165, 250, 0.24), transparent 65%);
  opacity: 0.85;
  z-index: 0;
  pointer-events: none;
}

.luxe-metric__icon {
  font-size: 1.15rem;
  opacity: 0.7;
}

.luxe-metric__label {
  font-size: 0.78rem;
  letter-spacing: 0.12em;
  text-transform: uppercase;
  opacity: 0.72;
}

.luxe-metric__value {
  font-size: 1.48rem;
  font-weight: 700;
}

.luxe-metric__delta {
  font-size: 0.82rem;
  opacity: 0.75;
}

.luxe-metric__caption {
  font-size: 0.82rem;
  color: var(--luxe-muted);
}

.luxe-stack {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(var(--stack-min, 16rem), 1fr));
  gap: var(--stack-gap, 1.1rem);
  margin: var(--stack-margin, 1.6rem 0 0);
}

.luxe-card {
  position: relative;
  border-radius: 22px;
  border: 1px solid var(--luxe-border);
  background: rgba(12, 17, 27, 0.72);
  padding: var(--card-padding, 1.3rem 1.4rem);
  box-shadow: 0 18px 40px rgba(8, 15, 35, 0.35);
  overflow: hidden;
  backdrop-filter: blur(18px);
  color: var(--luxe-ink);
}

.luxe-card::before {
  content: "";
  position: absolute;
  inset: 0;
  background: linear-gradient(135deg, rgba(96, 165, 250, 0.14), transparent 55%);
  opacity: 0.75;
  pointer-events: none;
}

.luxe-card__icon {
  font-size: 1.4rem;
  margin-bottom: 0.4rem;
}

.luxe-card__title {
  font-size: 1.05rem;
  margin: 0 0 0.45rem 0;
}

.luxe-card__body {
  font-size: 0.92rem;
  color: var(--luxe-muted);
}

.luxe-card__footer {
  margin-top: 0.7rem;
  font-size: 0.8rem;
  color: rgba(226, 232, 240, 0.64);
}
</style>
"""


def _inject_css() -> None:
  """Inject the shared CSS rules once per session."""
  if st.session_state.get("_luxe_css_injected"):
      return
  st.markdown(_LUXE_CSS, unsafe_allow_html=True)
  st.session_state["_luxe_css_injected"] = True


# ---------------------------------------------------------------------------
# Shared helpers
# ---------------------------------------------------------------------------
def _merge_styles(base: Mapping[str, str], extra: Mapping[str, str] | None = None) -> str:
  merged: dict[str, str] = dict(base)
  if extra:
      merged.update({k: v for k, v in extra.items() if v is not None})
  return "; ".join(f"{k}: {v}" for k, v in merged.items())


def ChipRow(
    chips: Sequence[str | Mapping[str, str]],
    *,
    tone: str | None = None,
    size: str = "md",
    render: bool = True,
    gap: str | None = None,
) -> str:
  """Render a reusable pill row.

  Parameters
  ----------
  chips:
      Sequence with either plain strings or dicts containing ``label`` and
      optional ``icon``/``tone`` overrides.
  tone:
      Base tone applied to chips (``accent``, ``info``, ``positive``...).
  size:
      ``"sm"``, ``"md"`` or ``"lg"``; controls padding and font size.
  render:
      When ``True`` the HTML is pushed to Streamlit, otherwise only returned.
  gap:
      Optional override for the spacing between chips.
  """
  _inject_css()

  size_map: Mapping[str, tuple[str, str]] = {
      "sm": ("0.32rem 0.8rem", "0.74rem"),
      "md": ("0.38rem 0.9rem", "0.82rem"),
      "lg": ("0.45rem 1.05rem", "0.92rem"),
  }
  padding, font_size = size_map.get(size, size_map["md"])
  row_style = {
      "--chip-padding": padding,
      "--chip-size": font_size,
  }
  if gap:
      row_style["--chip-gap"] = gap

  html = [f"<div class='luxe-chip-row' style='{_merge_styles(row_style, {})}'>"]
  for item in chips:
      if isinstance(item, Mapping):
          label = item.get("label", "")
          icon = item.get("icon")
          chip_tone = item.get("tone", tone)
      else:
          label = str(item)
          icon = None
          chip_tone = tone
      icon_fragment = f"<span>{icon}</span>" if icon else ""
      html.append(
          f"<span class='luxe-chip' data-tone='{chip_tone or ''}'>"
          f"{icon_fragment}<span>{label}</span>"
          "</span>"
      )
  html.append("</div>")
  html_markup = "".join(html)
  if render:
      st.markdown(html_markup, unsafe_allow_html=True)
  return html_markup


# ---------------------------------------------------------------------------
# TeslaHero
# ---------------------------------------------------------------------------
@dataclass
class TeslaHero:
  title: str
  subtitle: str
  chips: Sequence[str | Mapping[str, str]] = field(default_factory=list)
  icon: str | None = None
  gradient: str | None = None
  glow: str | None = None
  density: str = "cozy"
  parallax_icons: Sequence[Mapping[str, str]] = field(default_factory=list)

  def render(self) -> None:
      _inject_css()
      padding_map = {
          "compact": "1.9rem 2.2rem",
          "cozy": "2.5rem 2.9rem",
          "roomy": "3.1rem 3.4rem",
      }
      padding = padding_map.get(self.density, padding_map["cozy"])
      hero_style = {
          "--hero-padding": padding,
      }
      if self.gradient:
          hero_style["--hero-gradient"] = self.gradient
      if self.glow:
          hero_style["--hero-glow"] = self.glow

      layers = []
      for idx, layer in enumerate(self.parallax_icons):
          icon = layer.get("icon", "âœ¦")
          top = layer.get("top", f"{10 + idx * 12}%")
          left = layer.get("left", f"{55 + idx * 8}%")
          size = layer.get("size", "4rem")
          speed = layer.get("speed", f"{16 + idx * 4}s")
          layers.append(
              f"<span class='luxe-hero__layer' style='top:{top};left:{left};--layer-size:{size};--layer-speed:{speed};'>"
              f"{icon}</span>"
          )

      chips_html = ChipRow(self.chips, render=False) if self.chips else ""
      icon_html = f"<div class='luxe-hero__icon'>{self.icon}</div>" if self.icon else ""

      html = f"""
      <div class='luxe-hero' style='{_merge_styles(hero_style, {})}'>
        {''.join(layers)}
        <div class='luxe-hero__content'>
          {icon_html}
          <h1>{self.title}</h1>
          <p>{self.subtitle}</p>
          {chips_html}
        </div>
      </div>
      """
      st.markdown(html, unsafe_allow_html=True)


# ---------------------------------------------------------------------------
# MetricGalaxy
# ---------------------------------------------------------------------------
@dataclass
class MetricItem:
  label: str
  value: str
  caption: str | None = None
  delta: str | None = None
  icon: str | None = None
  tone: str | None = None


@dataclass
class MetricGalaxy:
  metrics: Sequence[MetricItem]
  glow: bool = True
  density: str = "cozy"
  min_width: str = "13rem"

  def render(self) -> None:
      _inject_css()
      padding_map = {
          "compact": "1rem 1.15rem",
          "cozy": "1.2rem 1.4rem",
          "roomy": "1.45rem 1.65rem",
      }
      gap_map = {
          "compact": "0.8rem",
          "cozy": "1rem",
          "roomy": "1.3rem",
      }
      style = {
          "--metric-padding": padding_map.get(self.density, padding_map["cozy"]),
          "--metric-gap": gap_map.get(self.density, gap_map["cozy"]),
          "--metric-min": self.min_width,
      }
      html = [f"<div class='luxe-metric-galaxy' style='{_merge_styles(style, {})}'>"]
      for metric in self.metrics:
          tone = metric.tone or ("positive" if (metric.delta and metric.delta.startswith("+")) else "")
          icon_html = f"<div class='luxe-metric__icon'>{metric.icon}</div>" if metric.icon else ""
          delta_html = f"<div class='luxe-metric__delta'>{metric.delta}</div>" if metric.delta else ""
          caption_html = f"<div class='luxe-metric__caption'>{metric.caption}</div>" if metric.caption else ""
          html.append(
              f"<div class='luxe-metric' data-glow='{str(self.glow).lower()}' data-tone='{tone}'>"
              f"{icon_html}"
              f"<div class='luxe-metric__label'>{metric.label}</div>"
              f"<div class='luxe-metric__value'>{metric.value}</div>"
              f"{delta_html}"
              f"{caption_html}"
              "</div>"
          )
      html.append("</div>")
      st.markdown("".join(html), unsafe_allow_html=True)


# ---------------------------------------------------------------------------
# GlassStack
# ---------------------------------------------------------------------------
@dataclass
class GlassCard:
  title: str
  body: str
  icon: str | None = None
  footer: str | None = None


@dataclass
class GlassStack:
  cards: Sequence[GlassCard]
  columns_min: str = "16rem"
  density: str = "cozy"

  def render(self) -> None:
      _inject_css()
      padding_map = {
          "compact": "1.05rem 1.15rem",
          "cozy": "1.3rem 1.4rem",
          "roomy": "1.6rem 1.75rem",
      }
      gap_map = {
          "compact": "0.9rem",
          "cozy": "1.1rem",
          "roomy": "1.45rem",
      }
      style = {
          "--card-padding": padding_map.get(self.density, padding_map["cozy"]),
          "--stack-gap": gap_map.get(self.density, gap_map["cozy"]),
          "--stack-min": self.columns_min,
      }
      html = [f"<div class='luxe-stack' style='{_merge_styles(style, {})}'>"]
      for card in self.cards:
          icon_html = f"<div class='luxe-card__icon'>{card.icon}</div>" if card.icon else ""
          footer_html = f"<div class='luxe-card__footer'>{card.footer}</div>" if card.footer else ""
          html.append(
              f"<div class='luxe-card'>"
              f"{icon_html}"
              f"<h3 class='luxe-card__title'>{card.title}</h3>"
              f"<div class='luxe-card__body'>{card.body}</div>"
              f"{footer_html}"
              "</div>"
          )
      html.append("</div>")
      st.markdown("".join(html), unsafe_allow_html=True)


__all__ = [
  "TeslaHero",
  "MetricGalaxy",
  "MetricItem",
  "GlassStack",
  "GlassCard",
  "ChipRow",
]
