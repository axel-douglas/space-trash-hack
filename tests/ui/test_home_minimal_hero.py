from pathlib import Path

from app.modules.luxe_components import HeroFlowStage, MinimalHero


def test_home_minimal_hero_snapshot() -> None:
    flow = [
        HeroFlowStage(
            key="inventory",
            order=1,
            name="Inventario",
            hero_headline="Calibr√° el inventario",
            hero_copy="Normaliz√° residuos, detect√° flags EVA y estructuras multi-layer.",
            card_body="Normaliz√° residuos y marc√° flags problem√°ticos (multilayer, EVA, nitrilo).",
            icon="üß±",
            timeline_label="Inventario en vivo",
            timeline_description="Inger√≠ CSV NASA, normaliz√° unidades y marca riesgos EVA desde la cabina.",
            footer="Dataset NASA + crew flags",
        ),
        HeroFlowStage(
            key="target",
            order=2,
            name="Target",
            hero_headline="Seleccion√° objetivo",
            hero_copy="Define l√≠mites de agua, energ√≠a y log√≠stica con presets marcianos.",
            card_body="Eleg√≠ producto final y l√≠mites de agua, energ√≠a y crew para la misi√≥n.",
            icon="üéØ",
            timeline_label="Target marciano",
            timeline_description="Seleccion√° producto final, l√≠mites de agua y energ√≠a, o usa presets homologados.",
            footer="Presets o l√≠mites manuales",
        ),
    ]

    metrics = [
        {
            "key": "status",
            "label": "Estado",
            "value": "‚úÖ Modelo listo",
            "caption": "Nombre: rexai-rf-ensemble",
            "icon": "üõ∞Ô∏è",
            "stage_key": "inventory",
            "tone": "accent",
        },
        {
            "key": "training",
            "label": "Entrenamiento",
            "value": "12 May 2045 10:00 UTC",
            "caption": "Procedencia: NASA ¬∑ Muestras: 128",
            "icon": "üß™",
            "stage_key": "target",
            "tone": "info",
        },
        {
            "key": "uncertainty",
            "label": "Incertidumbre",
            "value": "¬±4.2%",
            "caption": "Bandas 95% expuestas",
            "icon": "üìà",
            "stage_key": "results",
        },
    ]

    hero = MinimalHero(
        title="Rex-AI orquesta el reciclaje orbital y marciano",
        subtitle=(
            "Un loop aut√≥nomo que mezcla regolito MGS-1, pol√≠meros EVA y residuos de carga "
            "para fabricar piezas listas para misi√≥n."
        ),
        icon="üõ∞Ô∏è",
        chips=[
            {"label": "RandomForest multisalida", "tone": "accent"},
            {"label": "Comparadores heur√≠sticos", "tone": "info"},
            {"label": "Crew telemetry ready", "tone": "accent"},
        ],
        gradient="linear-gradient(135deg, rgba(59,130,246,0.28), rgba(14,165,233,0.08))",
        glow="rgba(96,165,250,0.45)",
        density="roomy",
        metrics=metrics,
        flow=flow,
    )

    markup = hero.markup.strip()
    snapshot_path = Path(__file__).with_name("__snapshots__") / "test_home_minimal_hero.py.snap"
    expected = snapshot_path.read_text(encoding="utf-8").strip()

    assert markup == expected
    assert markup.count("class='luxe-hero luxe-hero--minimal'") == 1
    assert markup.count("class='luxe-hero__kpi'") == 2
    assert "luxe-hero__layer" not in markup
    assert "video" not in markup
