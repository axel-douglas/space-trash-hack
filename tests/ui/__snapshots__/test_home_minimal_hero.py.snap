<div class='luxe-hero luxe-hero--minimal' style='--hero-padding: 3.1rem 3.4rem; --hero-gradient: linear-gradient(135deg, rgba(59,130,246,0.28), rgba(14,165,233,0.08)); --hero-glow: rgba(96,165,250,0.45)'>
          <div class='luxe-hero__content'>
            <div class='luxe-hero__icon'>🛰️</div>
            <h1>Rex-AI orquesta el reciclaje orbital y marciano</h1>
            <p>Un loop autónomo que mezcla regolito MGS-1, polímeros EVA y residuos de carga para fabricar piezas listas para misión.</p>
            <div class='luxe-chip-row' style='--chip-padding: 0.38rem 0.9rem; --chip-size: 0.82rem'><span class='luxe-chip' data-tone='accent'><span>RandomForest multisalida</span></span><span class='luxe-chip' data-tone='info'><span>Comparadores heurísticos</span></span><span class='luxe-chip' data-tone='accent'><span>Crew telemetry ready</span></span></div>
            <div class='luxe-hero__kpis'><article class='luxe-hero__kpi' data-tone='accent'><span class='luxe-hero__kpi-icon'>🛰️</span><div><span class='luxe-hero__kpi-value'>✅ Modelo listo</span><span class='luxe-hero__kpi-label'>Estado</span><span class='luxe-hero__kpi-caption'>Nombre: rexai-rf-ensemble</span></div></article><article class='luxe-hero__kpi' data-tone='info'><span class='luxe-hero__kpi-icon'>🧪</span><div><span class='luxe-hero__kpi-value'>12 May 2045 10:00 UTC</span><span class='luxe-hero__kpi-label'>Entrenamiento</span><span class='luxe-hero__kpi-caption'>Procedencia: NASA · Muestras: 128</span></div></article></div>
          </div>
        </div>
